packages/angular_devkit/build_angular/src/webpack/plugins/devtools-ignore-plugin.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { Compilation, Compiler } from 'webpack';
+
+// Following the naming conventions from
+// https://sourcemaps.info/spec.html#h.ghqpj1ytqjbm
+const IGNORE_LIST = 'x_google_devtoolsIgnore';
+
+const PLUGIN_NAME = 'devtools-ignore-plugin';
+
+interface Options {
+  vendors?: false | { test: RegExp };
+  runtime?: false | { test: RegExp };
+}
+
+interface SourceMap {
+  sources: string[];
+  [IGNORE_LIST]: number[];
+}
+
+/**
+ * This plugin adds a field to source maps that identifies which sources are
+ * vendored or runtime-injected (aka third-party) sources. These are consumed by
+ * Chrome DevTools to automatically ignore-list sources.
+ */
+export class DevToolsIgnorePlugin {
+  constructor(private options: Options) {}
+
+  apply(compiler: Compiler) {
+    const { RawSource } = compiler.webpack.sources;
+
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tap(
+        {
+          name: PLUGIN_NAME,
+          stage: Compilation.PROCESS_ASSETS_STAGE_DEV_TOOLING,
+          additionalAssets: true,
+        },
+        (assets) => {
+          let vendorsTest: RegExp | null = null;
+          let runtimeTest: RegExp | null = null;
+          if (this.options.vendors) {
+            vendorsTest = this.options.vendors.test;
+          }
+          if (this.options.runtime) {
+            runtimeTest = this.options.runtime.test;
+          }
+          if (!vendorsTest && !runtimeTest) {
+            return;
+          }
+
+          for (const [name, asset] of Object.entries(assets)) {
+            // Instead of using `asset.map` we process the map files RawSource.
+            // This is because `.map()` is slow and take several seconds.
+            if (!name.endsWith('.map')) {
+              // Ignore non map files
+              continue;
+            }
+
+            const mapContent = asset.source().toString();
+            if (!mapContent) {
+              continue;
+            }
+
+            const map = JSON.parse(mapContent) as SourceMap;
+
+            map[IGNORE_LIST] = Object.entries(map.sources)
+              .filter(([, source]) => vendorsTest?.test(source) || runtimeTest?.test(source))
+              .map(([index]) => +index);
```

Comment:
To avoid the conversion, this could be changed to something similar to:
```
map[IGNORE_LIST] = map.sources
        .map((source, index) => [source, index] as const)
        .filter(([source]) => vendorsTest?.test(source) || runtimeTest?.test(source))
        .map(([, index]) => index);
```
Other alternatives could be using `reduce` or a `for` loop.

Reply:
@dgp1130 Object.entries gives string keys for arrays. Hence the `+` to convert to number. `Number(index)` is also fine.

Reply:
Switched to a for loop.